reasoner/resolution/framework/Resolver.java

```diff
@@ -30,107 +29,59 @@
 public abstract class Resolver<T extends Resolver<T>> extends Actor.State<T> {
     private static final Logger LOG = LoggerFactory.getLogger(Resolver.class);
 
-    protected final String name;
-    private final Map<Request, ResponseProducer> responseProducers;
+    private final String name;
     private final Map<Request, Request> requestRouter;
-    private boolean isInitialised;
+    protected final ResolverRegistry registry;
 
-    public Resolver(Actor<T> self, String name) {
+    protected Resolver(Actor<T> self, String name, ResolverRegistry registry) {
         super(self);
         this.name = name;
-        isInitialised = false;
-        responseProducers = new HashMap<>();
-        requestRouter = new HashMap<>();
+        this.registry = registry;
+        this.requestRouter = new HashMap<>();
+        // Note: initialising downstream actors in constructor will create all actors ahead of time, so it is non-lazy
+        // additionally, it can cause deadlock within ResolverRegistry as different threads initialise actors
     }
 
     public String name() {
         return name;
     }
 
-    protected abstract ResponseProducer createResponseProducer(Request fromUpstream);
+    public abstract void receiveRequest(Request fromUpstream, int iteration);
 
-    protected abstract void initialiseDownstreamActors(ResolverRegistry registry);
+    protected abstract void receiveAnswer(Response.Answer fromDownstream, int iteration);
 
-    protected abstract Either<Request, Response> receiveRequest(Request fromUpstream, ResponseProducer responseProducer);
+    protected abstract void receiveExhausted(Response.Exhausted fromDownstream, int iteration);
 
-    protected abstract Either<Request, Response> receiveAnswer(Request fromUpstream, Response.Answer fromDownstream, ResponseProducer responseProducer);
+    protected abstract void initialiseDownstreamActors();
 
-    protected abstract Either<Request, Response> receiveExhausted(Request fromUpstream, Response.Exhausted fromDownstream, ResponseProducer responseProducer);
+    protected abstract ResponseProducer responseProducerCreate(Request fromUpstream, int iteration);
 
-    /*
-     *
-     * Handlers for messages sent into the execution actor that are dispatched via the actor model.
-     *
-     */
-    public void executeReceiveRequest(Request fromUpstream, ResolverRegistry registry) {
```

Comment:
the "meat" of these methods, which is actually not that much, has been moved down into each actor. This was mostly to facilitate the `RootResolver`, which does not require maintaining a `Map<Request, ResponseProducer>`, nor does it have to return a `Response` to the parent (it does not have a parent!)